controllers/machine_controller_node.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	apicorev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controllers/noderefutil"
+	capierrors "sigs.k8s.io/cluster-api/errors"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrNodeNotFound = errors.New("cannot find node with matching ProviderID")
+)
+
+func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	logger := r.Log.WithValues("machine", machine.Name, "namespace", machine.Namespace)
+	// Check that the Machine hasn't been deleted or in the process.
+	if !machine.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	logger = logger.WithValues("cluster", cluster.Name)
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		logger.Info("Machine doesn't have a valid ProviderID yet")
+		return nil
+	}
+
+	providerID, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
+	if err != nil {
+		return err
+	}
+
+	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		return err
+	}
+
+	// Even if Status.NodeRef exists, continue to do the following checks to make sure Node is healthy
+	node, err := r.getNode(remoteClient, providerID)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, getting ErrNodeNotFound error means the node is deleted.
+			// If Status.NodeRef is not set before, node still can be in the provisioning state.
+			if machine.Status.NodeRef != nil {
+				conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeNotFoundReason, clusterv1.ConditionSeverityError, "")
+			}
+			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 20 * time.Second},
```

Comment:
Should we do the same if ProviderID is missing?

Reply:
Actually, once ProviderID is there and we cannot get the node due to `ErrNodeNotFound`, it must have already passed provisioning. Because here `getNode()` is using providerID.

Reply:
Opened this PR for cleaning up RequeueAfterError so that we can remove it here too: https://github.com/kubernetes-sigs/cluster-api/pull/3723